[clangd] Drop the optimization where only shards for modified files are updated in the background index

HighCommander4 opened the pull request:
The optimization is not valid, given that the target of a symbol reference can change without the spelling of the reference (or anything else in the file containing the reference) changing, as illustrated in the added test case.

Partially fixes https://github.com/clangd/clangd/issues/1104

HighCommander4 committed:
[clangd] Drop the optimization where only shards for modified files are updated in the background index

The optimization is not valid, given that the target of a symbol
reference can change without the spelling of the reference (or
anything else in the file containing the reference) changing,
as illustrated in the added test case.

Partially fixes https://github.com/clangd/clangd/issues/1104

kadircet commented:
I am afraid this doesn't address clangd/clangd#1104 to a useful extent :/.

Whenever a header `foo.h` changes, clangd will schedule indexing of only a **single** translation unit that depends on it, not all of them. So if there are other TUs that reference `foo`, they'll still stay stale.

This was a deliberate trade-off back in the day, for performance wins, there are other cases where files contents can stay the same but USRs for symbols inside the file, and the ones being referenced might change (command line flags, different PP states, ...) and we assumed they're "rare" enough and will be compensated by dynamic index until user modifies these other files.

I guess this assumption didn't age well, I think your idea around having a way to flush/rebuild background-index is probably most straight forward approach. We didn't have one explicitly, as there's always the workaround of deleting `$PROJECT_ROOT/.cache/clangd/index/`. As I believe the performance trade-offs here are overwhelming and it'll result in pretty much re-indexing of whole project at every rebase or slight modifications to headers (we can't satisfy correctness without rebuilding all the TUs that depend on a changed header, rather than a single one).

HighCommander4 committed:
Add a second testcase illustrating a remaining problem

Out of multiple source files including a header, only one
is re-indexed when the header is modified.

HighCommander4 commented:
> I am afraid this doesn't address [clangd/clangd#1104](https://github.com/clangd/clangd/issues/1104) to a useful extent :/.
> 
> Whenever a header `foo.h` changes, clangd will schedule indexing of only a **single** translation unit that depends on it, not all of them. So if there are other TUs that reference `foo`, they'll still stay stale.

Good catch, thanks; I had overlooked that.

(I added an additional testcase demonstrating this, mostly for my own future reference / as a reference for anyone else who might pick this up.)

> I believe the performance trade-offs here are overwhelming and it'll result in pretty much re-indexing of whole project at every rebase or slight modifications to headers (we can't satisfy correctness without rebuilding all the TUs that depend on a changed header, rather than a single one).

FWIW, my expectation as a user is that when a header is modified, that **does** trigger re-indexing of all source files that include the header. I don't think the performance impact of this is that catastrophic, as the number of headers in a project tends to vary inversely with how widely they're included (i.e. most headers will have a few includers, and a few will have many). In my mind, the scenario is comparable to modifying a header and then **rebuilding**: if the header is widely included, you will rebuild many TUs, and programmers expect that (and adopt practices like forward declarations, pimpl, etc. to avoid unnecessary header dependencies).

Anyways, I agree that the patch in its current form is not useful. I'll mark it as a draft, and (when time permits) will re-work it to implement a "proper" (all dependent TUs rebuilt) update procedure. We can then discuss what the trigger for that procedure should be (explicit invocation, clangd restart, file-save of header, etc.).

JVApen commented:
> FWIW, my expectation as a user is that when a header is modified, that **does** trigger re-indexing of all source files that include the header. I don't think the performance impact of this is that catastrophic, as the number of headers in a project tends to vary inversely with how widely they're included (i.e. most headers will have a few includers, and a few will have many). In my mind, the scenario is comparable to modifying a header and then **rebuilding**: if the header is widely included, you will rebuild many TUs, and programmers expect that (and adopt practices like forward declarations, pimpl, etc. to avoid unnecessary header dependencies).

At ours, indexing the whole project is measured is hours. There are a couple of headers that are included a lot, like the header that suppresses compiler warnings. I don't think it adds value to reindex everything when such a commonly inserted header is touched.

I also don't believe that you can compare reindexing with recompilation. We do distributed compilation only on remote machines, such that your local machine is not unresponsive when you do touch that kind of headers. Assuming indexing and compilation take the same time, a 10x speedup would for a build of 10 minutes give you 1h40m of indexing. (Without any other changes)

As such, I think it would be reasonable to fallback to the current behavior if the header touches more than X files. Once X is large enough, I don't think that it adds a lot of value, as you don't want to wait for 1h for it to catch up in order to make use of it.
